### ISSUE TITLE
Testing accuracy always more than 99%. 

### ISSUE THREAD

**Issue by rudraptpsingh** [fchollet/keras#1878]
I am trying to implement a language model using LSTMs. My network runs fine and I also see that the training loss decreases but the testing accuracy is always above 99% even if I don not train my network for long. 
I have used word2vec vectors and embedded the weights in the Embedding layer. My network looks like:

model = Graph()
model.add_input(name='input', input_shape=(n_train,), dtype=int)
model.add_node(Embedding(output_dim=rnn_dim, input_dim=n_symbols, weights=[embedding_weights]),name = 'embedding',input='input') 
model.add_node(LSTM(output_dim=dense_dim,input_dim=rnn_dim), name='forward', input='embedding')
model.add_node(Dropout(0.5), name='dropout', input='forward')
model.add_node(Dense(output_size, activation='softmax'), name='softmax', input='dropout')
model.add_output(name='output', input='softmax')

sgd = SGD(lr=0.01, decay=1e-6, momentum=0.99, nesterov=True)
model.compile(sgd, loss={'output': 'categorical_crossentropy'})
print("Train...")
model.fit({'input': X_train,'output': y_train},
          batch_size=128,
          nb_epoch=1,verbose=1)

My training and testing array shapes are:
X_train shape: (100000, 18)
X_test shape: (10000, 18)
y_train shape: (100000, 998)
y_test shape: (10000, 998)

where there are 100000 training and 10000 testing sentences and each sentence contains 18 words. The number of classes for output is 998. 

Can anyone suggest why I might not be getting a genuine testing error?

**Comment by eralmansouri** [comment]
Try removing the line "model.add_output(~)". I believe you only need to use that when you are combining outputs of more than 1 previous layers. 

In your situation I think it's assuming 99% accuracy because atleast 996/998 of the outputs are correct.

**Comment by rudraptpsingh** [comment]
I think the network works fine when I changed 
model.add_node(Dense(output_size, activation='softmax'), name='softmax', input='dropout')
to 
model.add_node(Dense(output_size, activation='sigmoid), name='softmax', input='dropout')

But I am still unclear on this and need more intuition.

**Comment by philipperemy** [comment]
What do you want to say when you think that your network works? You think it's too suspicious to have 99% accuracy? Sigmoid and Softmax are different concepts. Sigmoid maps a real value to [0,1] which can be thought of being an absolute probability. Softmax takes all the values and computes the relative probability (of one word compared to all the other categories). Softmax is used for classification purposes.

When you use already pre-trained weights in your word2vec, it would make sense that the model would converge much quicker than if they are initialized to random values. 

By the way, are your sentences in a context or independent?

**Comment by parag2489** [comment]
Assuming the code snippet you have posted is the complete one... I don't see any calculation of testing accuracy. Is your "training" accuracy 99% instead of testing?

**Comment by rudraptpsingh** [comment]
Ok. So found the issue which is genuine. For multiple classes, the activation used should be softmax. I used that but I am seeing a very slow training with a slowly increasing accuracy and decreasing loss. The test/train accuracy are somewhere around 25% and the loss around 5. I think this should decrease much more. For a given batch of training sequence, I am using 25 epochs. Should I increase this or decrease this? Also, I am using SGD with learning rate 0.01, weight decay 10e-6 and Nesterov momentum=0.90. I am also using tanh activation in my LSTM layer. Should I add/change something in order for the network to perform better?

**Comment by philipperemy** [comment]
More epochs is better. Make sure that the validation loss does not increase after a while. To prevent overfitting, you can add an early stopping condition:

from http://keras.io/callbacks/
```
keras.callbacks.EarlyStopping(monitor='val_loss', patience=0, verbose=0, mode='auto')
```
Using the categorical cross entropy is a good choice since this is the default loss function for multi-class classification problems and softmax output units. You can modify the learning rate to 0.1 and add a learning rate decay.

Try to tune your layers hyperparameters such as the number of cells you have in your LSTM, Dense and Dropout layers.
